File: pkg/models/large_language_model.go
package models

// RecognizedReceiptImageResponse represents a view-object of recognized receipt image response
type RecognizedReceiptImageResponse struct {
	Type                 TransactionType `json:"type"`
	Time                 int64           `json:"time,omitempty"`
	CategoryId           int64           `json:"categoryId,string,omitempty"`
	SourceAccountId      int64           `json:"sourceAccountId,string,omitempty"`
	DestinationAccountId int64           `json:"destinationAccountId,string,omitempty"`
	SourceAmount         int64           `json:"sourceAmount,omitempty"`
	DestinationAmount    int64           `json:"destinationAmount,omitempty"`
	TagIds               []string        `json:"tagIds,omitempty"`
	Comment              string          `json:"comment,omitempty"`
}

// RecognizedReceiptImageResult represents the result of recognized receipt image
type RecognizedReceiptImageResult struct {
	Type                   string   `json:"type,omitempty" jsonschema:"enum=income,enum=expense,enum=transfer" jsonschema_description:"Transaction type (income, expense, transfer)"`
	Time                   string   `json:"time" jsonschema:"format=date-time" jsonschema_description:"Transaction time in long date time format (YYYY-MM-DD HH:mm:ss, e.g. 2023-01-01 12:00:00)"`
	Amount                 string   `json:"amount,omitempty" jsonschema_description:"Transaction amount"`
	AccountName            string   `json:"account,omitempty" jsonschema_description:"Account name for the transaction"`
	CategoryName           string   `json:"category,omitempty" jsonschema_description:"Category name for the transaction"`
	TagNames               []string `json:"tags,omitempty" jsonschema_description:"List of tags associated with the transaction (maximum 10 tags allowed)"`
	Description            string   `json:"description,omitempty" jsonschema_description:"Transaction description"`
	DestinationAmount      string   `json:"destination_amount,omitempty" jsonschema_description:"Destination amount for transfer transactions"`
	DestinationAccountName string   `json:"destination_account,omitempty" jsonschema_description:"Destination account name for transfer transactions"`
}

const (
	AIAssistantModeChat    = "chat"
	AIAssistantModeSummary = "summary"
)

// AIAssistantHistoryItem represents one history message for ai assistant
type AIAssistantHistoryItem struct {
	Role    string `json:"role" binding:"required,oneof=user assistant"`
	Content string `json:"content" binding:"required,max=2048"`
}

// AIAssistantChatRequest represents all parameters for ai assistant chat request
type AIAssistantChatRequest struct {
	Mode    string                    `json:"mode" binding:"omitempty,oneof=chat summary"`
	Message string                    `json:"message" binding:"max=2048"`
	History []*AIAssistantHistoryItem `json:"history" binding:"max=20"`
}

// AIAssistantReferencedTransaction represents one referenced transaction in ai assistant response
type AIAssistantReferencedTransaction struct {
	Id                     int64           `json:"id,string"`
	Time                   int64           `json:"time"`
	TimeText               string          `json:"timeText,omitempty"`
	Type                   TransactionType `json:"type"`
	CategoryName           string          `json:"categoryName,omitempty"`
	SourceAccountName      string          `json:"sourceAccountName,omitempty"`
	DestinationAccountName string          `json:"destinationAccountName,omitempty"`
	SourceAmount           int64           `json:"sourceAmount"`
	DestinationAmount      int64           `json:"destinationAmount,omitempty"`
	Currency               string          `json:"currency,omitempty"`
	DestinationCurrency    string          `json:"destinationCurrency,omitempty"`
	Comment                string          `json:"comment,omitempty"`
	SimilarityScore        float64         `json:"similarityScore,omitempty"`
}

// AIAssistantChatResponse represents ai assistant chat response
type AIAssistantChatResponse struct {
	Mode       string                              `json:"mode"`
	Reply      string                              `json:"reply"`
	References []*AIAssistantReferencedTransaction `json:"references,omitempty"`
}

// AIAssistantResult represents the result schema of ai assistant response from llm
type AIAssistantResult struct {
	Reply string `json:"reply,omitempty" jsonschema_description:"Response text for user with bill summary and bookkeeping suggestions"`
}

File: src/models/large_language_model.ts
export interface RecognizedReceiptImageResponse {
    readonly type: number;
    readonly time?: number;
    readonly categoryId?: string;
    readonly sourceAccountId?: string;
    readonly destinationAccountId?: string;
    readonly sourceAmount?: number;
    readonly destinationAmount?: number;
    readonly tagIds?: string[];
    readonly comment?: string;
}

export type AIAssistantMode = 'chat' | 'summary';
export type AIAssistantMessageRole = 'user' | 'assistant';

export interface AIAssistantHistoryItem {
    readonly role: AIAssistantMessageRole;
    readonly content: string;
}

export interface AIAssistantChatRequest {
    readonly mode: AIAssistantMode;
    readonly message?: string;
    readonly history?: AIAssistantHistoryItem[];
}

export interface AIAssistantReferencedTransaction {
    readonly id: string;
    readonly time: number;
    readonly timeText?: string;
    readonly type: number;
    readonly categoryName?: string;
    readonly sourceAccountName?: string;
    readonly destinationAccountName?: string;
    readonly sourceAmount: number;
    readonly destinationAmount?: number;
    readonly currency?: string;
    readonly destinationCurrency?: string;
    readonly comment?: string;
    readonly similarityScore?: number;
}

export interface AIAssistantChatResponse {
    readonly mode: AIAssistantMode;
    readonly reply: string;
    readonly references?: AIAssistantReferencedTransaction[];
}

export type AIAssistantChatStreamChunkType = 'thinking_delta' | 'reply_delta' | 'references' | 'done';

export interface AIAssistantChatStreamChunk {
    readonly type: AIAssistantChatStreamChunkType;
    readonly mode?: AIAssistantMode;
    readonly delta?: string;
    readonly reply?: string;
    readonly thinking?: string;
    readonly references?: AIAssistantReferencedTransaction[];
}
